dex/pair/src/safe_price.rs

```diff
@@ -91,188 +42,108 @@ pub trait SafePriceModule:
-        future: CumulativeState<Self::Api>,
-    ) {
-        if !current.is_default() {
-            self.current_state().set(&current);
-        }
-        if !future.is_default() {
-            self.future_state().set(&future);
+        if price_observations.len() == safe_price_params.max_observations {
+            let _ = price_observations.set(new_index, &new_price_observation);
+        } else {
+            price_observations.push(new_price_observation);
         }
-    }
+        safe_price_params.current_index = new_index;
 
-    fn has_max_observations(&self, current_state: &CumulativeState<Self::Api>) -> bool {
-        let max_observations_per_record = self.get_max_observations_per_record();
-        current_state.num_observations == max_observations_per_record
+        price_observations_mapper.set(&price_observations);
+        safe_price_params_mapper.set(&safe_price_params);
     }
 
-    fn has_half_max_observations(&self, current_state: &CumulativeState<Self::Api>) -> bool {
-        let max_observations_per_record = self.get_max_observations_per_record();
-        current_state.num_observations == max_observations_per_record / 2
-    }
-
-    fn get_current_state_or_default(&self) -> CumulativeState<Self::Api> {
-        if !self.current_state().is_empty() {
-            self.current_state().get()
+    fn compute_new_observation(
+        &self,
+        new_round: Round,
+        new_first_reserve: &BigUint,
+        new_second_reserve: &BigUint,
+        current_price_observation: &PriceObservation<Self::Api>,
+    ) -> PriceObservation<Self::Api> {
+        let new_weight = if current_price_observation.recording_round > 0 {
+            new_round - current_price_observation.recording_round
         } else {
-            Default::default()
-        }
-    }
+            1
+        };
 
-    fn get_future_state_or_default(&self) -> CumulativeState<Self::Api> {
-        if !self.future_state().is_empty() {
-            self.future_state().get()
-        } else {
-            Default::default()
-        }
+        // Create a new variable, to avoid overwriting the old price observation
+        let mut new_price_observation = current_price_observation.clone();
+        new_price_observation.first_token_reserve_accumulated +=
+            BigUint::from(new_weight) * new_first_reserve;
+        new_price_observation.second_token_reserve_accumulated +=
+            BigUint::from(new_weight) * new_second_reserve;
+        new_price_observation.weight_accumulated += new_weight;
+        new_price_observation.recording_round = new_round;
+
+        new_price_observation
     }
 
-    fn get_max_observations_per_record(&self) -> u64 {
-        if !self.max_observations_per_record().is_empty() {
-            self.max_observations_per_record().get()
+    #[endpoint(setSafePriceParams)]
+    fn set_safe_price_params(
+        &self,
+        new_max_observations: usize,
+        default_safe_price_rounds_offset: u64,
```

Comment:
I think this should be in the view contract only. No need to add it into the safePriceParams.

Reply:
Updated.